backend/dataall/modules/shares_base/db/share_object_state_machines.py

```diff
@@ -0,0 +1,296 @@
+import logging
+
+from dataall.base.db import exceptions
+from dataall.modules.shares_base.services.shares_enums import (
+    ShareObjectActions,
+    ShareObjectStatus,
+    ShareItemActions,
+    ShareItemStatus,
+)
+from dataall.modules.shares_base.db.share_object_repositories import ShareObjectRepository
+
+logger = logging.getLogger(__name__)
+
+
+class Transition:
+    def __init__(self, name, transitions):
+        self._name = name
+        self._transitions = transitions
+        self._all_source_states = [*set([item for sublist in transitions.values() for item in sublist])]
```

Comment:
nits
* `itertools.chain.from_iterable(transitions.values())` should flatten your values
*  why convert the set to list? I guess you want to dedup but why convert back to list?
*  by converting to set you assume that the values are hashable, I can see that are enum values so they must be but give it a thought